### PR TITLE
Infra/core/move header bit manipulation to utility detail

### PR DESCRIPTION
**Pull request by Irallia**
This PR is part of https://github.com/seqan/product_backlog/issues/160 and resolves the part of the ToDos for the top-level header bit_manipulation.hpp together with #2292 .

The task was to split bit_manipulation.hpp into utility/detail/to_little_endian and utility/detail/bits_of (plus renaming of sizeof_bits into bits_of)
and the remaining stuff goes into seqan3/std/bit which is done by #2292

### CODE REVIEW

**Review by MitraDarja — APPROVED** [review]
Looks good to me. :)

File: include/seqan3/contrib/stream/bgzf_stream_util.hpp [hunk 1]
```diff
@@ -13,8 +13,10 @@
 #pragma once
 
 #include <array>
+#include <seqan3/std/algorithm>
```
Comment (rrahn) [inline include/seqan3/contrib/stream/bgzf_stream_util.hpp:16]:
```suggestion
#include <seqan3/std/algorithm>
#include <array>
```

File: include/seqan3/utility/detail/bits_of.hpp [hunk 2]
```diff
@@ -118,7 +109,7 @@ constexpr uint8_t popcount(unsigned_t const n) noexcept
 }
 
 /*!\brief Returns the number of leading 0-bits, starting at the most significant bit position.
-* \ingroup core
+*
```
Comment (rrahn) [inline include/seqan3/utility/detail/bits_of.hpp:112]:
```suggestion
 * \ingroup utility
```
?

**Review by rrahn — CHANGES_REQUESTED** [review]

**Comment by smehringer** [comment]
@Irallia You can request core again, if the PR is ready

**Comment by eseiler** [comment]
Is this not blocked by https://github.com/seqan/seqan3/pull/2292 anymore?
I'd rather not move all the stuff that is deprecated in #2292, i.e. just move the stuff to the new header that will actually stay there.
This also has the advantage that you only need to remove the old file and not change the new one again. (and also less merge conflicts).
We did it like this for `include/seqan3/alignment/band/static_band.hpp` and `include/seqan3/core/detail/endian.hpp`.

**Comment by eseiler** [comment]
I think we should merge #2292 first and then rebase this.

**Comment by smehringer** [comment]
This can be rebased now

File: test/performance/utility/detail/bits_of_benchmark.cpp [hunk 3]
```diff
@@ -51,7 +51,7 @@ static void is_power_of_two_seqan3(benchmark::State & state) {
     for (auto _ : state)
     {
         n = std::rand();
-        benchmark::DoNotOptimize(seqan3::detail::is_power_of_two(n));
+        benchmark::DoNotOptimize(std::has_single_bit(n));
```
Comment (marehr) [inline test/performance/utility/detail/bits_of_benchmark.cpp:54]: please revert:

```suggestion
        benchmark::DoNotOptimize(seqan3::detail::is_power_of_two(n));
```

File: test/performance/utility/detail/bits_of_benchmark.cpp [hunk 4]
```diff
@@ -76,7 +76,7 @@ static void next_power_of_two_seqan3(benchmark::State & state) {
     for (auto _ : state)
     {
         n = std::rand();
-        benchmark::DoNotOptimize(seqan3::detail::next_power_of_two(n));
+        benchmark::DoNotOptimize(std::bit_ceil(n));
```
Comment (marehr) [inline test/performance/utility/detail/bits_of_benchmark.cpp:79]: please revert:

```suggestion
        benchmark::DoNotOptimize(seqan3::detail::next_power_of_two(n));
```

File: test/performance/utility/detail/bits_of_benchmark.cpp [hunk 5]
```diff
@@ -9,7 +9,7 @@
```
Comment (marehr) [inline test/performance/utility/detail/bits_of_benchmark.cpp:9]: The file should be called `seqan3/test/performance/std/bit_benchmark.hpp`, can you move it there?

File: test/unit/utility/detail/bits_of_test.cpp [hunk 6]
```diff
@@ -115,31 +67,29 @@ TYPED_TEST(unsigned_operations, most_significant_bit_set)
     }
 }
 
-TYPED_TEST(unsigned_operations, count_leading_zeros)
+TYPED_TEST(unsigned_operations, countl_zero)
```
Comment (marehr) [inline test/unit/utility/detail/bits_of_test.cpp:70]: This test should be moved to `std/bit_test.cpp`

File: test/unit/utility/detail/bits_of_test.cpp [hunk 7]
```diff
@@ -83,7 +35,7 @@ class unsigned_operations : public ::testing::Test
 
 TYPED_TEST_SUITE(unsigned_operations, unsigned_types, );
 
-TYPED_TEST(unsigned_operations, most_significant_bit_set)
+TYPED_TEST(unsigned_operations, bit_width)
```
Comment (marehr) [inline test/unit/utility/detail/bits_of_test.cpp:38]: This test should be moved to `std/bit_test.cpp`

File: test/unit/utility/detail/bits_of_test.cpp [hunk 8]
```diff
@@ -152,22 +102,22 @@ TYPED_TEST(unsigned_operations, count_leading_zeros)
 TYPED_TEST(unsigned_operations, count_trailing_zeros)
```
Comment (marehr) [inline test/unit/utility/detail/bits_of_test.cpp:102]:
```suggestion
TYPED_TEST(unsigned_operations, countr_zero)
```

Comment (marehr) [reply test/unit/utility/detail/bits_of_test.cpp:102]: This test should be moved to `std/bit_test.cpp`

**Review by marehr — COMMENTED** [review]
Hi thank you! This is a first sweep.

I tried to go commit-by-commit, but noticed that your history isn't clean and some rebase's weren't resolved correctly.

File: test/performance/utility/detail/bits_of_benchmark.cpp (hunk 3 above)
Comment (Irallia) [reply test/performance/utility/detail/bits_of_benchmark.cpp]: Why? I thougt `is_power_of_two` and `bit_ceil` is now deprecated? And the whole file `bit_manipulation.hpp` aswell?

**Review by marehr — APPROVED** [review]

**Pull request merged**